text/2020-03-11-versioned-kv.md

```diff
@@ -0,0 +1,210 @@
+  it is difficult to solve the problems of online data migration, incremental
+   backup, and data synchronization between two data centers because timestamp
+  is not supported.
+
+In order to solve the above problems, this proposal presents another kv
+interface - VerKV, which mainly addsversion information such as timestamp to
+the key. At the same time, this interface will be designed to be approximate
+to the semantics of HBase, which will facilitate the service migration from
+HBase to TiKV, thereby integrating TiKV better into the HBase ecosystem.
+
+## Detailed designs
+
+### Functions and scenarios
+
+Here are the functionality designs and the corresponding application
+scenarios for VerKV:
+
+- Support data migration, incremental backup, CDC (Change Data Capture)
+  and cross-IDC cluster synchronization application scenarios include:
+
+  - Online data replication between TiKV clusters, which includes a full
+    migrations followed by incremental migrations;
+  - Cluster data backup, which includes a first full backup followed perform
+    an incremental data backups every other time;
+  - CDC service created based on incremental data acquisition; can reliably
+    provide incremental data services;
+  - In cross data center cluster synchronization scenarios, VerKV can not
+    only process incremental data synchronization, but also handle write
+    conflicts caused by the same key.
+
+- Support multi-versioned data. The corresponding application scenarios
+  include:
+
+  - Data version retrieval. For certain businesses, multiple versions of
+    historical data must be maintained for for backtracking and analysis.
+    Reliance on third-party storage and consumption of computing resources
+    (user dynamic characteristics, User behavior link tracking) needs to be
+    reduced, and the data processing process optimized;
+  - Data version control, for fast data recovery after problematic data
+    writes in businesses.
+
+- Support data expiration based on TTL. The corresponding application
+  scenarios include:
+
+  - Setting expiration time to obtain data of a recent period
+  - Setting expiration time to clear inactive data for a specified amount
+    of time
+
+### Compatibility
+
+The design adds a new Column Family called `VerCF` in RocksDB, which avoids
+potential compatibility issues with the existing RawKV annd TxnKV data and
+processing logics, and could be created dynamically when needed.
+This design allows maximum reuse of the current implementations of TiKV and PD.
+
+### Data model and interface
+
+VerKV is a kv-like interface similar to the RawKV interface. It hides version
+information from upper-layer services. Its data model is multi-versioned KV
+between RawKV and TxnKV.
+
+#### Data model and Consistency
+
+The data model is like `key:version -> value`. The version uses the current
+timestamp by default, or can be specified by the user (such as the global `ts`
```

Comment:
In either case, I would prefer the client to supply the timestamp. If TiKV knows the current time, that is adding a chunk of new functionality/complexity. In fact, it would be really nice if timestamps could remain opaque to TiKV as far as possible (which I think means we can't support the TTL expiry functionality in TiKV).

Reply:
I would prefer client giving timestamp as well.